### PR TITLE
[4.0] Download key manager - GSoC Expand Extensions Manager Redo

### PR DESCRIPTION
**Pull request by roland-d**
### Redo
This is a redo of #17575 as I can no longer push to the original repo.

### Introduction
Currently there is no clear overview of which extensions have download keys to make users download their extensions with an extra_query with their keys included to allow the download. Developers have been currently hacking this system by creating their own view to add a download key to the extra_query field in the #__update_sites table (Example of an extra_query field: dlid=`key`&amp;dummy=my.zip).
Each extension developer may have a different way to build this extra query so the idea is to make a view to manage and insert the download keys of all extensions in a single view in com_installer.

To set or customize the prefix and suffix of the download key you need to have an <dlid> tag with prefix and suffix in your installation file like this:

`<dlid prefix="dlid=" suffix="&amp;dummy=my.zip"/>`
(note: remember to close />)

If an extension has no download ID set in the manifest file, the download key will show up as N/A in the list and will not be editable.

### Testing Instructions

Here are the extensions to test the feature. I have not included all types as I think these 3 already make clear what this is about.
1. Run the [MySQL update query](https://github.com/joomla/joomla-cms/pull/25553/files#diff-adc3eedd33560141a69c0cb8d1d4214e) or the [Postgres update query](https://github.com/joomla/joomla-cms/pull/25553/files#diff-7b5d5176f5c5de2280a23dcd12f034ea) depending on the database you use
2. Apply the patch
3. Install the extensions below
Component
[keycomponent.zip](https://github.com/joomla/joomla-cms/files/3390080/keycomponent.zip)
Plugin:
[keyplugin.zip](https://github.com/joomla/joomla-cms/files/3390082/keyplugin.zip)
Module:
[keymodule.zip](https://github.com/joomla/joomla-cms/files/3390083/keymodule.zip)
4. Go to System -> Update Sites
image

The list shows just the name of the extension and the update site URL.

5. Clicking on the extension name will open the Edit screen.
image
5. Enter a download Key
6. Click on Save & Close
7. The download key will now be shown in the list
image

The Download Key is shown without the prefix and suffix. Only the developer will be responsible to define the prefix and suffix of the download key.

You are going to need a real extension with a download key to see if the download key is send. For now you could also check the database by doing this:
1. Go to the table __update_sites table
2. Check the field `extra_query` for the extension where you added the download key. It should have the format of `prefix` `key` `suffix`

### Expected result

A simple a intuitive way to the user manage his download keys and less work to the developers to create a view to generate the extra_query

### Actual result
There is no single place to manage your download keys.

### Documentation Changes Required
Yes, once merged I will write a document on how to incorporate the tags in the installer XML

### CODE REVIEW

File: administrator/components/com_installer/View/Downloadkey/HtmlView.php [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+
+		if (empty($this->item->get('update_site_id')))
+		{
+			$toolbar->cancel('downloadkey.cancel');
+		}
+		else
+		{
+			if (ComponentHelper::isEnabled('com_contenthistory')
+				&& $this->state->params->get('save_history', 0)
+				&& $canDo->get('core.edit')
```
Comment (Quy) [inline administrator/components/com_installer/View/Downloadkey/HtmlView.php]:
```suggestion
				&& $canDo->get('core.edit'))
```

File: administrator/components/com_installer/View/Downloadkey/HtmlView.php [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+			)
```
Comment (Quy) [inline administrator/components/com_installer/View/Downloadkey/HtmlView.php]:
```suggestion
```

File: administrator/language/en-GB/en-GB.mod_menu.ini [hunk 3]
```diff
@@ -98,11 +98,13 @@ MOD_MENU_INSTALLER_SUBMENU_DATABASE="Database"
 MOD_MENU_INSTALLER_SUBMENU_MANAGE="Manage"
 MOD_MENU_INSTALLER_SUBMENU_UPDATE="Update"
 MOD_MENU_INSTALLER_SUBMENU_UPDATESITES="Update Sites"
+MOD_MENU_INSTALLER_SUBMENU_DOWNLOADKEYS="Download Keys Manager"
```
Comment (Quy) [inline administrator/language/en-GB/en-GB.mod_menu.ini]: Alpha sort

File: administrator/language/en-GB/en-GB.ini [hunk 4]
```diff
@@ -665,6 +666,7 @@ JHELP_COMPONENTS_COM_CONTENT_OPTIONS="Components_Article_Manager_Options"
 JHELP_COMPONENTS_COM_CSP_OPTIONS="Components_CSP_Reports_Options"
 JHELP_COMPONENTS_COM_FINDER_OPTIONS="Components_Smart_Search_Configuration"
 JHELP_COMPONENTS_COM_INSTALLER_OPTIONS="Components_Installer_Configuration"
+JHELP_COMPONENTS_COM_INSTALLER_DOWNLOADKEYS_EDIT="Components_Installer_Downloadkeys_Edit"
```
Comment (Quy) [inline administrator/language/en-GB/en-GB.ini:672]: Alpha sort

**Comment by mbabker** [comment]
I still don't believe this is a core responsibility.  Has anyone gauged interest among the extension ecosystem to determine how many vendors would provide support for this instead of their own solutions?  Because without ecosystem support, you're going to have a interface in core that end users think they are going to be able to use but the reality of it is if extensions don't support it it's just going to confuse the heck out of users wondering why they can't use a core feature.

These types of projects come across as core looking for problems to solve without figuring out if it is a real problem to be solved (or moreso someone not liking the fact that extension responsibilities are left to the extension to be implemented).

**Comment by joeforjoomla** [comment]
Agree with @mbabker 
I don't think this feature is widely used in the Joomla ecosystem.
Additionally, having such a feature in the core would be even more confusing for end users and developers.

**Comment by rdeutz** [comment]
I can see a value in this, we have build a complete extension install from our backend and extensions make Joomla great even if you can do a lot with core.

**Comment by MastersOfMedia** [comment]
The points raised by @mbabker and @joeforjoomla are, with all due respect, part of a chicken/egg discussion. If we want to improve the handling of download/update keys, and thus updates themselves (which is much needed imho), something needs to be done. The current situation where we leave this to developers is confusing and time-consuming to both end-users and developers. In my own experience, a lot of my customers leave extensions unpatched/not updated because they can not be updated with the built-in updater. This is an undesired effect of the current situation, unwanted for all of us.

If such a feature will become part of the Joomla! core, and we could persuade at least a number of developers to participate and make use of it, this could be a very strong feature and make Joomal! even more user-friendly 

Erik

**Comment by joeforjoomla** [comment]
@MastersOfMedia with all due respect for you, imho i see this as something optional given that only a small subset of paid extension use it and that must be managed by the extension itself when and if needed, not by the Joomla core.

**Comment by mbabker** [comment]
> In my own experience, a lot of my customers leave extensions unpatched/not updated because they can not be updated with the built-in updater. This is an undesired effect of the current situation, unwanted for all of us.

This pull request is not going to fix that.  It's still up to the developer to build support into their extensions to use the updater, just like it's still going to be up to the developer to build support into their extensions to use a core "owned" download key manager.  Who says this key manager is going to work with all of the systems that developers are presently using for their extensions, or that developers are going to commit to rewriting their extensions (and possibly their backend systems validating things) to use this system when they already have a working system (basically, why make change for the sake of change).

Is it useful?  Yes.  Is this core territory?  IMO, no.  And if it were core territory, there would have been more feature requests asking for it over the years, unless I've had my head buried in the sand this only came up within the GSoC team the year the project was accepted (which is part of my reason for the "core looking for problems to solve" comment before).

**Comment by cpaschen** [comment]
I agree with everyone ... we need it, it shouldn't be in the core, its a chicken/egg thing.
I think the best solution would be for one of the major extension developers (akeeba/regular labs) to develop this as an extension first that all can use. THEN once it has lots of user and developers using it (which would only happen if one of those MAJOR extension devs release it) it could get to the point where other developers use it and THEN it might be wise to have it in the core.
But until something like that happens, I agree, it might be more 'excess core never used' feature (and I agree, might be more confusing to users).
So ... anyone else willing to pick this up and run with it?

File: administrator/components/com_installer/Controller/DownloadkeyController.php [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+ */
+
+namespace Joomla\Component\Installer\Administrator\Controller;
+
+defined('_JEXEC') or die;
+
+use Joomla\CMS\MVC\Controller\FormController;
+
+/**
+ * Controller for a single contact
```
Comment (wilsonge) [inline administrator/components/com_installer/Controller/DownloadkeyController.php]: not a contact

File: administrator/components/com_installer/Helper/InstallerHelper.php [hunk 6]
```diff
@@ -14,6 +14,7 @@
 use Joomla\CMS\Factory;
 use Joomla\CMS\HTML\HTMLHelper;
 use Joomla\CMS\Language\Text;
+use stdClass;
```
Comment (wilsonge) [inline administrator/components/com_installer/Helper/InstallerHelper.php]: We're not importing global namespace things so far and I'd rather keep it that way for now please. Let's just strap a `\` on the front of the instance for now please

**Comment by wilsonge** [comment]
@cpaschen the download key system (which has been around for a while just without a UI) already came from akeeba (i don't know if nic still does use this as he decouples his stuff more and more from Joomla Core). However there's no central place to manage this. I believe this does have a place in core as the questions on how to use the download key in modules and plugins (and there was no easy way). So I believe this is actually fine for Joomla as long as extensions opt into it in their XML files (so only extensions that use the Joomla Download Key mechanism show up in the list of extensions to add a key to)

File: administrator/components/com_installer/access.xml [hunk 7]
```diff
@@ -4,7 +4,7 @@
 		<action name="core.admin" title="JACTION_ADMIN" />
 		<action name="core.options" title="JACTION_OPTIONS" />
 		<action name="core.manage" title="JACTION_MANAGE" />
-		<action name="core.delete" title="JACTION_DELETE" />
+		<action name="core.edit" title="JACTION_EDIT" />
```
Comment (wilsonge) [inline administrator/components/com_installer/access.xml:8]: Isn't `core.delete` used for uninstalling extensions?

Comment (roland-d) [reply administrator/components/com_installer/access.xml:8]: It was deleted by accident.

File: administrator/components/com_installer/tmpl/downloadkeys/default.php [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+										</a>
+									<?php else: ?>
+										<?php echo $item->update_site_name; ?>
+									<?php endif; ?>
+									<br>
+									<span class="small break-word">
+										<a href="<?php echo $item->location; ?>" target="_blank" rel="noopener noreferrer"><?php echo $this->escape($item->location); ?></a>
+									</span>
+								</th>
+								<td class="hidden-sm-down text-center">
```
Comment (brianteeman) [inline administrator/components/com_installer/tmpl/downloadkeys/default.php]: Should be left aligned like the header row
```suggestion
								<td class="hidden-sm-down">
```

File: administrator/components/com_installer/tmpl/downloadkeys/default.php [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+									<span class="bold hasTooltip" title="<?php echo HTMLHelper::_('tooltipText', $item->name, $item->description, 0); ?>">
+										<?php echo $item->name; ?>
+									</span>
+								</td>
+								<td class="hidden-sm-down text-center">
```
Comment (brianteeman) [inline administrator/components/com_installer/tmpl/downloadkeys/default.php]: Should be left aligned like the header row
```suggestion
								<td class="hidden-sm-down">
```

File: administrator/components/com_installer/tmpl/downloadkeys/default.php [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+									<?php echo $item->client_translated; ?>
+								</td>
+								<td class="hidden-sm-down text-center">
```
Comment (brianteeman) [inline administrator/components/com_installer/tmpl/downloadkeys/default.php]: Should be left aligned like the header row
```suggestion
								<td class="hidden-sm-down">
```

File: administrator/components/com_installer/tmpl/downloadkeys/default.php [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+									<?php echo $item->type_translated; ?>
+								</td>
+								<td class="hidden-sm-down text-center">
```
Comment (brianteeman) [inline administrator/components/com_installer/tmpl/downloadkeys/default.php]: Should be left aligned like the header row
```suggestion
								<td class="hidden-sm-down">
```

File: administrator/components/com_installer/tmpl/downloadkeys/default.php [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+									<?php echo $item->folder_translated; ?>
+								</td>
+								<td class="hidden-sm-down text-center">
```
Comment (brianteeman) [inline administrator/components/com_installer/tmpl/downloadkeys/default.php]: Should be left aligned like the header row
```suggestion
								<td class="hidden-sm-down">
```

File: administrator/components/com_installer/tmpl/downloadkeys/default.php [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+$listDirn  = $this->escape($this->state->get('list.direction'));
+?>
+<div id="installer-manage" class="clearfix">
+	<form action="<?php echo Route::_('index.php?option=com_installer&view=downloadkeys'); ?>" method="post" name="adminForm" id="adminForm">
+		<div class="row">
+			<div class="col-md-12">
+				<div id="j-main-container" class="j-main-container">
+					<?php echo JLayoutHelper::render('joomla.searchtools.default', ['view' => $this]); ?>
+					<?php if (empty($this->items)) : ?>
+					<div class="alert alert-warning alert-no-items">
```
Comment (brianteeman) [inline administrator/components/com_installer/tmpl/downloadkeys/default.php]:
```suggestion
					<div class="alert alert-info">
```

File: administrator/components/com_installer/tmpl/downloadkeys/default.php [hunk 14]
```diff
@@ -0,0 +1,122 @@
...
+						<?php echo Text::_('JGLOBAL_NO_MATCHING_RESULTS'); ?>
```
Comment (brianteeman) [inline administrator/components/com_installer/tmpl/downloadkeys/default.php]:
```suggestion
						<span class="fa fa-info-circle" aria-hidden="true"></span><span class="sr-only"><?php echo Text::_('INFO'); ?></span>						
						<?php echo Text::_('JGLOBAL_NO_MATCHING_RESULTS'); ?>
```

**Comment by brianteeman** [comment]
Please look at how I added a `<caption>` to tables 
https://github.com/joomla/joomla-cms/blob/dab646292fb38fd15b2fb336207ea28d6f1d8e51/administrator/components/com_contact/tmpl/contacts/default.php#L49-L51

File: administrator/components/com_installer/forms/downloadkey.xml [hunk 15]
```diff
@@ -0,0 +1,25 @@
...
+        <field
+                name="update_site_id"
+                type="hidden"
+        />
+        <field
+                name="extra_query"
+                type="text"
+                label="COM_INSTALLER_DOWNLOADKEY_EXTRA_QUERY_LABEL"
+                size="40"
+                class="input-xxlarge"
```
Comment (brianteeman) [inline administrator/components/com_installer/forms/downloadkey.xml]:
```suggestion
```

We dont need this class in j4

**Comment by brianteeman** [comment]
Is there a reason that this is seperate from com_installer&view=updatesites

File: administrator/components/com_installer/tmpl/downloadkeys/default.php [hunk 16]
```diff
@@ -0,0 +1,122 @@
...
+					</div>
+					<?php else : ?>
+					<table class="table table-striped">
```
Comment (brianteeman) [inline administrator/components/com_installer/tmpl/downloadkeys/default.php]:
```suggestion
					<table class="table">
```

File: administrator/components/com_installer/forms/filter_downloadkeys.xml [hunk 17]
```diff
@@ -0,0 +1,78 @@
...
+<form addfieldprefix="Joomla\Component\Installer\Administrator\Field">
+    <fields name="filter">
+        <field
+                name="search"
+                type="text"
+                label="COM_INSTALLER_UPDATESITES_FILTER_SEARCH_LABEL"
+                description="COM_INSTALLER_UPDATESITES_FILTER_SEARCH_DESC"
+                hint="JSEARCH_FILTER"
+        />
+        <field
```
Comment (brianteeman) [inline administrator/components/com_installer/forms/filter_downloadkeys.xml]: there is nothing to filter by here as the rows do not have a state so just delete this field

File: administrator/components/com_installer/forms/downloadkey.xml (hunk 15 above)
Comment (brianteeman) [inline administrator/components/com_installer/forms/downloadkey.xml]:
```suggestion
```

File: administrator/components/com_installer/forms/filter_downloadkeys.xml [hunk 18]
```diff
@@ -0,0 +1,78 @@
+<?xml version="1.0" encoding="utf-8"?>
+<form addfieldprefix="Joomla\Component\Installer\Administrator\Field">
+    <fields name="filter">
+        <field
+                name="search"
+                type="text"
+                label="COM_INSTALLER_UPDATESITES_FILTER_SEARCH_LABEL"
```
Comment (brianteeman) [inline administrator/components/com_installer/forms/filter_downloadkeys.xml]: this label and desc should be unique for _downloadkeys and not re-using _updatesites

also see my question about combining the two

File: administrator/components/com_installer/forms/filter_downloadkeys.xml [hunk 19]
```diff
@@ -0,0 +1,78 @@
...
+                onchange="this.form.submit();"
+                >
+                <option value="">JOPTION_SELECT_PUBLISHED</option>
+                <option value="0">JDISABLED</option>
+                <option value="1">JENABLED</option>
+        </field>
+        <field
+                name="client_id"
+                type="location"
+                onchange="this.form.submit();"
```
Comment (brianteeman) [inline administrator/components/com_installer/forms/filter_downloadkeys.xml]:
```suggestion
		label="COM_INSTALLER_VALUE_CLIENT_SELECT"
                onchange="this.form.submit();"
```

File: administrator/components/com_installer/forms/filter_downloadkeys.xml [hunk 20]
```diff
@@ -0,0 +1,78 @@
...
+                description="COM_PLUGINS_FILTER_PUBLISHED_DESC"
+                onchange="this.form.submit();"
+                >
+                <option value="">JOPTION_SELECT_PUBLISHED</option>
+                <option value="0">JDISABLED</option>
+                <option value="1">JENABLED</option>
+        </field>
+        <field
+                name="client_id"
+                type="location"
```
Comment (brianteeman) [inline administrator/components/com_installer/forms/filter_downloadkeys.xml]: You need to add labels for each field see this example
https://github.com/joomla/joomla-cms/blob/dab646292fb38fd15b2fb336207ea28d6f1d8e51/administrator/components/com_installer/forms/filter_updatesites.xml#L20-L26

File: administrator/components/com_installer/forms/filter_downloadkeys.xml [hunk 21]
```diff
@@ -0,0 +1,78 @@
...
+                >
+                <option value="">COM_INSTALLER_VALUE_FOLDER_SELECT</option>
+        </field>
+    </fields>
+    <fields name="list">
+        <field
+                name="fullordering"
+                type="list"
+                label="JGLOBAL_SORT_BY"
+                description="JGLOBAL_SORT_BY"
```
Comment (brianteeman) [inline administrator/components/com_installer/forms/filter_downloadkeys.xml]:
```suggestion
```

File: administrator/components/com_installer/forms/filter_downloadkeys.xml [hunk 22]
```diff
@@ -0,0 +1,78 @@
...
+                onchange="this.form.submit();"
+                default="name ASC"
+                >
+                <option value="">JGLOBAL_SORT_BY</option>
+                <option value="enabled ASC">JSTATUS_ASC</option>
```
Comment (brianteeman) [inline administrator/components/com_installer/forms/filter_downloadkeys.xml]:
```suggestion
```

File: administrator/components/com_installer/forms/filter_downloadkeys.xml [hunk 23]
```diff
@@ -0,0 +1,78 @@
...
+                <option value="enabled DESC">JSTATUS_DESC</option>
```
Comment (brianteeman) [inline administrator/components/com_installer/forms/filter_downloadkeys.xml]:
```suggestion
```

File: administrator/components/com_installer/forms/filter_downloadkeys.xml [hunk 24]
```diff
@@ -0,0 +1,78 @@
...
+                <option value="folder_translated ASC">COM_INSTALLER_HEADING_FOLDER_ASC</option>
+                <option value="folder_translated DESC">COM_INSTALLER_HEADING_FOLDER_DESC</option>
+                <option value="update_site_id ASC">JGRID_HEADING_ID_ASC</option>
+                <option value="update_site_id DESC">JGRID_HEADING_ID_DESC</option>
+        </field>
+        <field
+                name="limit"
+                type="limitbox"
+                label="JGLOBAL_LIMIT"
+                description="JGLOBAL_LIMIT"
```
Comment (brianteeman) [inline administrator/components/com_installer/forms/filter_downloadkeys.xml]:
```suggestion
```

File: administrator/components/com_installer/tmpl/downloadkey/default.xml [hunk 25]
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" encoding="utf-8"?>
+<metadata>
+	<layout title="Download Keys Manager">
```
Comment (brianteeman) [inline administrator/components/com_installer/tmpl/downloadkey/default.xml]: Should be a language string

File: administrator/components/com_installer/tmpl/downloadkey/default.xml [hunk 26]
```diff
@@ -0,0 +1,8 @@
...
+		<message>
+			<![CDATA[COM_INSTALLER_UPDATESITES_VIEW_DEFAULT_DESC]]>
```
Comment (brianteeman) [inline administrator/components/com_installer/tmpl/downloadkey/default.xml]: should be a _downloadkey string

File: administrator/components/com_installer/tmpl/downloadkeys/default.xml [hunk 27]
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" encoding="utf-8"?>
+<metadata>
+	<layout title="Download Keys Manager">
```
Comment (brianteeman) [inline administrator/components/com_installer/tmpl/downloadkeys/default.xml]: should be a language string

File: administrator/components/com_installer/tmpl/downloadkeys/default.xml [hunk 28]
```diff
@@ -0,0 +1,8 @@
...
+		<message>
+			<![CDATA[COM_INSTALLER_UPDATESITES_VIEW_DEFAULT_DESC]]>
```
Comment (brianteeman) [inline administrator/components/com_installer/tmpl/downloadkeys/default.xml]: should be a _downloadkey string

File: administrator/components/com_installer/tmpl/downloadkeys/default.php [hunk 29]
```diff
@@ -0,0 +1,123 @@
...
+ */
+
+defined('_JEXEC') or die;
+
+use Joomla\CMS\HTML\HTMLHelper;
+use Joomla\CMS\Language\Text;
+use Joomla\CMS\Router\Route;
+
+HTMLHelper::_('behavior.multiselect');
+HTMLHelper::_('bootstrap.tooltip');
```
Comment (brianteeman) [inline administrator/components/com_installer/tmpl/downloadkeys/default.php]:
```suggestion
```

Comment (brianteeman) [reply administrator/components/com_installer/tmpl/downloadkeys/default.php]: see comment below

File: administrator/components/com_installer/tmpl/downloadkeys/default.php [hunk 30]
```diff
@@ -0,0 +1,123 @@
...
+									<?php else: ?>
+										<?php echo $item->update_site_name; ?>
+									<?php endif; ?>
+									<br>
+									<span class="small break-word">
+										<a href="<?php echo $item->location; ?>" target="_blank" rel="noopener noreferrer"><?php echo $this->escape($item->location); ?></a>
+									</span>
+								</th>
+								<td class="hidden-sm-down">
+									<span class="bold hasTooltip" title="<?php echo HTMLHelper::_('tooltipText', $item->name, $item->description, 0); ?>">
```
Comment (brianteeman) [inline administrator/components/com_installer/tmpl/downloadkeys/default.php]:
```suggestion
									<div role="tooltip" id ="tip<?php echo $i; ?>">
										<?php echo $item->description; ?>
									</div>
```

File: administrator/components/com_installer/tmpl/downloadkeys/default.php [hunk 31]
```diff
@@ -0,0 +1,123 @@
...
+								</th>
+								<th scope="col" style="width:5%" class="d-none d-md-table-cell">
+									<?php echo HTMLHelper::_('searchtools.sort', 'JGRID_HEADING_ID', 'update_site_id', $listDirn, $listOrder); ?>
+								</th>
+								<th scope="col" style="width:5%" class="d-none d-md-table-cell">
+									<?php echo Text::_('COM_INSTALLER_HEADING_DOWNLOADKEY'); ?>
+								</th>
+							</tr>
+						</thead>
+						<tfoot>
```
Comment (brianteeman) [inline administrator/components/com_installer/tmpl/downloadkeys/default.php]: delete the entire tfoot and just put the getlistfooter line after the table see https://github.com/joomla/joomla-cms/blob/dab646292fb38fd15b2fb336207ea28d6f1d8e51/administrator/components/com_installer/tmpl/updatesites/default.php#L113-L114

File: administrator/components/com_plugins/View/Plugin/HtmlView.php [hunk 32]
```diff
@@ -112,5 +115,10 @@ protected function addToolbar()
 		}
 
 		ToolbarHelper::help($help->key, false, $url);
+
+		if (isset($this->item->xml->dlid))
+		{
+			ToolbarHelper::downloadkey('JGLOBAL_DOWNLOAD_KEY_MODAL_BUTTON', $this->item->extension_id);
```
Comment (brianteeman) [inline administrator/components/com_plugins/View/Plugin/HtmlView.php]: this is giving me the following error
0 Call to undefined method Joomla\CMS\Toolbar\ToolbarHelper::downloadkey()

File: administrator/language/en-GB/en-GB.ini [hunk 33]
```diff
@@ -339,6 +339,7 @@ JGLOBAL_DESCRIPTION="Description"
 JGLOBAL_DISPLAY_NUM="Display #"
 JGLOBAL_DISPLAY_SELECT_DESC="Show or hide the Display Select dropdown listbox."
 JGLOBAL_DISPLAY_SELECT_LABEL="Display Select"
+JGLOBAL_DOWN="Down"
```
Comment (brianteeman) [inline administrator/language/en-GB/en-GB.ini]: i dont see this being used anywhere

File: administrator/language/en-GB/en-GB.mod_menu.ini [hunk 34]
```diff
@@ -95,6 +95,7 @@ MOD_MENU_INSTALL_DISCOVER="Discover"
 MOD_MENU_INSTALL_EXTENSIONS="Extensions"
 MOD_MENU_INSTALL_LANGUAGES="Install Languages"
 MOD_MENU_INSTALLER_SUBMENU_DATABASE="Database"
+MOD_MENU_INSTALLER_SUBMENU_DOWNLOADKEYS="Download Keys Manager"
```
Comment (brianteeman) [inline administrator/language/en-GB/en-GB.mod_menu.ini]: we dont need this new string - all the submenu strings are not needed now and will be removed in another pr
```suggestion
```

File: libraries/src/Toolbar/ToolbarHelper.php [hunk 35]
```diff
@@ -39,9 +39,9 @@ abstract class ToolbarHelper
 	public static function title($title, $icon = 'generic.png')
 	{
 		$layout = new FileLayout('joomla.toolbar.title');
-		$html   = $layout->render(array('title' => $title, 'icon' => $icon));
+		$html   = $layout->render(['title' => $title, 'icon' => $icon]);
 
-		$app = Factory::getApplication();
+		$app                  = Factory::getApplication();
```
Comment (brianteeman) [inline libraries/src/Toolbar/ToolbarHelper.php]: ??

File: administrator/components/com_installer/forms/filter_downloadkeys.xml (hunk 17 above)
Comment (roland-d) [reply administrator/components/com_installer/forms/filter_downloadkeys.xml]: It actually does filter on the enabled/disabled state of the update site.

**Comment by roland-d** [comment]
@brianteeman 

> Is there a reason that this is seperate from com_installer&view=updatesites

None other than that nobody thought about it or asked about it. I do agree, it might be better integrated there. Especially since most of the data is all the same.

I am going to make the changes and integrate it there.

File: administrator/components/com_installer/forms/filter_downloadkeys.xml (hunk 17 above)
Comment (brianteeman) [reply administrator/components/com_installer/forms/filter_downloadkeys.xml]: Thats never going to be clear to a user ;)

But is probably another example of why this should probably be integrated with updatesources instead of being independent

**Comment by brianteeman** [comment]
> I am going to make the changes and integrate it there.

Great. It will make things much easier going forward. Less duplicate code to maintain etc etc

File: libraries/src/Toolbar/ToolbarHelper.php (hunk 35 above)
Comment (roland-d) [reply libraries/src/Toolbar/ToolbarHelper.php]: This is because the line below it is `$app->JComponentTitle = $html;` and this change aligns the `=`.

Comment (brianteeman) [reply libraries/src/Toolbar/ToolbarHelper.php]: ok

File: administrator/language/en-GB/en-GB.ini (hunk 33 above)
Comment (roland-d) [reply administrator/language/en-GB/en-GB.ini]: Removed.

File: administrator/components/com_plugins/View/Plugin/HtmlView.php (hunk 32 above)
Comment (roland-d) [reply administrator/components/com_plugins/View/Plugin/HtmlView.php]: I missed that. It should not be there so removed it.

File: administrator/components/com_installer/forms/filter_downloadkeys.xml (hunk 18 above)
Comment (roland-d) [reply administrator/components/com_installer/forms/filter_downloadkeys.xml]: This will be obsolete once it is integrated in the update sites

**Comment by roland-d** [comment]
@brianteeman How is this looking?

image

I placed the download key under the update site name and XML URL.

**Comment by brianteeman** [comment]
It would need some sort of label

**Comment by wilsonge** [comment]
How does the the update site list view work when an extension has more than 1 update site (like core for example). Is it multiple rows in that list (in which case download keys don't work) or is it multiple text boxes for a single extension?

**Comment by brianteeman** [comment]
Update site list view
image

Where are the multiples?

**Comment by roland-d** [comment]
@wilsonge The number of update sites is irrelevant to the key. A users download key (or extra_query as Joomla calls it) is still going to be the same. Per update site you can only enter 1 download key as this is how the database schema is.

**Comment by wilsonge** [comment]
@roland-d just to check this is still work in progress - because this PR seems pretty broken since your changes 3 days ago :)

**Comment by roland-d** [comment]
@wilsonge Yes I am aware it is broken. Hoping to fix it over the next few days. Today busy with my new site launch.

**Comment by wilsonge** [comment]
No stress. Just making sure :)

**Comment by wilsonge** [comment]
Is this ready for testing again?

**Comment by roland-d** [comment]
@wilsonge Yes, it is ready for testing again. I have updated all the instructions to reflect the latest changes.

Also pinging @sanderpotjer

**Comment by rdeutz** [comment]
@roland-d could you fix the phpcs stuff, thanks.

**Comment by roland-d** [comment]
@rdeutz I gave it a try, let's see if the system is happy now :)

**Comment by roland-d** [comment]
@rdeutz The errors now is not something I have a clue about how to fix and seem unrelated to this PR.

**Comment by bees4ever** [comment]
I have tested this item :white_check_mark: successfully on 8758172a007a15e55a12dc19d5e9f55429054c41<br /><br />I just compared the database table #__update_sites where in the field `extra_query` it is stored correctly. One thing mentionend: The example component has a spelling error where the xml file defining the "donwload key part" has a key `sufix`, but it should call `suffix` with double f. For some tester it could be useful to change that to see that the patch works correct.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25553">issues.joomla.org/tracker/joomla-cms/25553</a>.</sub>

**Comment by VladikB** [comment]
I have tested this item :white_check_mark: successfully on 8758172a007a15e55a12dc19d5e9f55429054c41<br /><br />As bees4ever wrote, I had to edit the keycomponent/downloadkeytestcomponent.xml .
<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25553">issues.joomla.org/tracker/joomla-cms/25553</a>.</sub>

**Comment by ghost** [comment]
Status "Ready To Commit". 1 build failing.

**Comment by wilsonge** [comment]
Thanks!

**Comment by roland-d** [comment]
Thank you thank you thank you

**Comment by roland-d** [comment]
Add documentation for this feature: https://docs.joomla.org/Manifest_files
